README.md
# @36node/shanghaibus-messenger

基于 kafka 的消息接收 clinet

[![version][0]][1] [![downloads][2]][3]

## Install

```bash
yarn add @36node/shanghaibus-messenger
```

## 使用

```js
import Messenger from "shanghaibus-messenger";

const client = new Messenger(
  {
    "group.id": "shanghaibus-core",
    "metadata.broker.list": "localhost:9092",
  },
  ["TBOX"]
);

/** 订阅全部 rdb log */
client.onRdbData(log => {
  console.log(log.vin, "rdb log comming");
});

/** 按照 command 订阅 rdb log */
client.onRdbData(
  log => {
    console.log(log.command);
  },
  ["REALTIME_REPORT", "REISSUE_REPORT"]
);

/**订阅 request error 类型log */
client.onRequestError(log => {
  console.log("Request error log comming");
});

/** 订阅 invalid log */
client.onInvalidLog(log => {
  console.log("Invalid log comming");
});
```

### messenger 解析后日志的基本格式

```typescript
{
  recordId?: string, // 由原始日志的 session + seq 组成， 对于 RDB_DATA 类型的日志才有，其他类型的日志没有这条记录
  cost?: number, // 解析日志的耗时
  data?: string, // 原始二进制日志内容(二进制字符串)
  origin?: string,  // 二进制字符串
  partial?: boolean, // 是否完全
  type: string, // 日志类型
  payload: object, // 日志的内容, 对于 REALTIME_REPORT, REISSUE_REPORT, payload 为 record
  vin?: string, // 车辆vin码， 只在 RDB_DATA 类型日志中存在
  command?: string, // 日志的命令， 只在 RDB_RATA 类型日志中存在
  error?: string, // 日志解析时发生的错误，只在 INVALID_LOG 类型中存在
  reportedAt: string, // 日志发送的时间，为 ISO string
  kafka_message: string, // kafka 传递过来的原始日志
}
```

### 日志类型(type)

| 类型          | 说明                         |
| :------------ | :--------------------------- |
| RDB_DATA      | 车辆信息日志                 |
| REQUEST_ERROR | 日志产生时发生的错误         |
| INVALID_ERROR | messenger 无法正确解析的日志 |

### 命令(command)

在 RDB_DATA 类型的日志中，按 command 可分为:

| 类型            | 说明         |
| :-------------- | :----------- |
| VEHICLE_LOGIN   | 车辆登入     |
| VEHICLE_LOGOUT  | 车辆登出     |
| HEARTBEAT       | 心跳         |
| REALTIME_REPORT | 实时数据上报 |
| REISSUE_REPORT  | 数据补发     |

### record 格式

将原始的车辆数据进行解析，形成 record

```typescript
{
  id:string, // 车辆vin
  at:string, // 信息采集时间
  vehicle?: object, // 整车数据, 具体数据内容见下原始日志格式说明
  montors?: [object], // 电机数据, 具体数据内容见下原始日志格式说明
  extreme?: object, // 极值数据, 具体数据内容见下原始日志格式说明
  location?: object, // 车辆位置信息，用 GCJ－02坐标系 表示
  alarm?: [string], // 警报数据, 为警报码的数组
  customExt?: object, // 自定义数据，具体数据内容见下原始日志格式说明
  adas?: [object], // 补充协议数据， item.type === "ADAS" 具体数据内容见下原始日志格式说明
  tens?: [object], // 补充协议数据，item.type === "TEN_SECONDS" 的数据
}
```

## 原始日志格式说明

原始日志指从 kafka 接收到的未经处理的日志

### 0 日志基本格式

```js
{
  level: number, // 日志级别
  time: number, // 日志产生时间，unix时间戳
  msg: string, // 日志信息 "handle rdb data", "request error", "session closed" 等
  request?: object, // 日志的信息，可选
  error?: object, // 错误信息, 可选
  session?: string, // 终端的session
  seq?: string, // 当前session下的消息编号
}
```

### 1 车辆实时信息日志(Handle rdb data)

#### 1.1 车辆登入 (command=VEHICLE_LOGIN)

```js
{
  level: 30, // 日志级别
  time: 1540902956033, // 日志时间戳
  msg: "handle rdb data", // 日志信息
  pid: 17,
  hostname: "d35202af7c8b-shanghaibus-v0-32960-1",
  session: "phJhOh3EiUF",
  seq: 2,
  cost: 0,
  origin:
    "232301fe4c5a595442474257364a3130313431393401001e120a1e14233600fd383938363034303231303137303031373937373901005c",
  request: { // 日志具体内容
    command: "VEHICLE_LOGIN",
    flag: "COMMAND",
    vin: "LZYTBGBW6J1014194", // 车辆vin码
    encrypt: "NONE",
    length: 30,
    body: {
      at: "2018-10-30T12:35:54.000Z", // 数据上报时间
      sn: 253,
      iccid: "89860402101700179779",
      subSysNm: 1,
      subSysNmLen: 0,
      subSysSn: [],
    },
  },
  response: "232301014c5a595442474257364a31303134313934010006120a1e1423364f",
  v: 1,
}

```

request body 格式说明:

| 字段        | 数据表示内容           | 数据类型        | 描述及要求                                                                          |
| :---------- | :--------------------- | :-------------- | :---------------------------------------------------------------------------------- |
| at          | 数据采集时间           | date ISO string | 数据采集的原始时间                                                                  |
| sn          | 登入流水号             | integer         | 车载终端每登入一次登入流水号自动加 1，从 1 开始循环累加，最大值 65531，循环周期为天 |
| iccid       | ICCID                  | string          | SIM 卡 ICCID 号（ICCID 应为终端从 SIM 卡获取的值，不应该人为填写或修改）            |
| subSysNm    | 可充电储能子系统数     | integer         | 可充电储能子系统数 n，有效值范围：0~250                                             |
| subSysNmLen | 可充电储能系统编码长度 | integer         | 可充电储能系统编码长度 m，有效值范围：0~50，“0”表示不上传该编码                     |
| subSysSn    | 可充电储能系统编码     | [string]        | 可充电储能系统编码宜为终端从车辆获取的值                                            |

#### 1.2 车辆登出 (command=VEHICLE_LOGOUT)

```js
{
  level: 30,
  time: 1540902976674,
  msg: "handle rdb data",
  pid: 17,
  hostname: "d35202af7c8b-shanghaibus-v0-32960-1",
  session: "fcG8xIaC4C8",
  seq: 470,
  cost: 0,
  origin: "232304fe4c53464430333230344a43303031353935010008120a1e1424110014e9",
  request: {
    command: "VEHICLE_LOGOUT",
    flag: "COMMAND",
    vin: "LSFD03204JC001595",
    encrypt: "NONE",
    length: 8,
    body: { at: "2018-10-30T12:36:17.000Z", sn: 20 },
  },
  response: "232304014c53464430333230344a43303031353935010006120a1e1424110c",
  v: 1,
}
```

request body 格式说明：

| 数据表示内容 | 字段 | 数据类型        | 描述及要求                     |
| :----------- | :--- | :-------------- | :----------------------------- |
| 登出时间     | at   | date ISO string | 数据采集的原始时间             |
| 登出流水号   | sn   | integer         | 登出流水号和当前登入流水号一致 |

#### 1.3 实时信息上报 (command=REALTIME_REPORT)

```js
{
  level: 30,
  time: 1540902953574,
  msg: "handle rdb data",
  pid: 17,
  hostname: "d35202af7c8b-shanghaibus-v0-32960-1",
  session: "z2-NYoZCXXc",
  seq: 749482,
  cost: 1,
  origin:
    "232302fe4c5a59544147425732453130353434393101007f120a1e142400010103010000001b390b164925da39023e41b6000102010104284e204e2028000027100500073d270801dcda6a06013f0cbf015b0cb2010246014e400700000000000000000080003000000003e803e8ffffffffffffffffffffffffffff1649feca00000000000000000000ffff00000000ff280028282802c7",
  request: {
    command: "REALTIME_REPORT",
    flag: "COMMAND",
    vin: "LZYTAGBW2E1054491",
    encrypt: "NONE",
    length: 127,
    body: {
      at: "2018-10-30T12:36:00.000Z",
      items: [
        {
          type: "VEHICLE",
          status: "ON",
          chargeStatus: "UNCHARGED",
          mode: "ELECTRIC",
          speed: 0,
          mileage: 178407.5,
          voltage: 570.5,
          current: -31,
          soc: 0.57,
          dcStatus: "OFF",
          shift: "D",
          resistance: 16822,
          aptv: 0,
          brake: 0.01,
        },
        {
          type: "MOTOR",
          count: 1,
          motors: [
            {
              no: 1,
              status: "READY",
              controlTemp: 0,
              speed: 0,
              torque: 0,
              temp: 0,
              voltage: 0,
              current: 0,
            },
          ],
        },
        { type: "LOCATION", state: 0, lng: 121.4482, lat: 31.25105 },
        {
          type: "EXTREME",
          maxVoltageSubSysNo: 1,
          maxVoltageSingNo: 63,
          maxVoltage: 3.263,
          minVoltageSubSysNo: 1,
          minVoltageSingNo: 91,
          minVoltage: 3.25,
          maxNtcSubSysNo: 1,
          maxNtcNo: 2,
          maxNtc: 30,
          minNtcSubSysNo: 1,
          minNtcNo: 78,
          minNtc: 24,
        },
        {
          type: "ALARM",
          maxLevel: 0,
          uas: {
            ressChargeOver: 0,
            motorTemp: 0,
            highVolMuteStatus: 0,
            motorControlTemp: 0,
            dcdcStatus: 0,
            brake: 0,
            dcdcTemp: 0,
            insulation: 0,
            batteryBadConsistency: 0,
            ressNotMatch: 0,
            socJump: 0,
            socOver: 0,
            batteryLow: 0,
            batteryOver: 0,
            socLow: 0,
            ressVolLow: 0,
            ressVolOver: 0,
            batteryTempOver: 0,
            tempDiff: 0,
          },
          ressLen: 0,
          ressList: [],
          mortorLen: 0,
          mortorList: [],
          engineLen: 0,
          engineList: [],
          otherLen: 0,
          otherList: [],
        },
        {
          type: "CUSTOM_EXT",
          dataLen: 48,
          pressure1: 0,
          pressure2: 0,
          batteryVoltage: 0,
          dcov: -900,
          dcoc: -900,
          cv: 570.5,
          rc: 5522.6,
          cp: 0,
          totalCharge: 0,
          totalDischarge: 0,
          bpiRes: 0,
          bniRes: 0,
          motorContTemp: 0,
          airMode: "OFF",
          airTemp: 0,
          insideTemp: 0,
          outsideTemp: 0,
          middleDoorStatus: "CLOSE",
          frontDoorStatus: "CLOSE",
          handbrakeStatus: "OFF",
          keyPosition: "ON",
        },
      ],
    },
  },
  response: "232302014c5a595441474257324531303534343931010006120a1e14240071",
  v: 1,
}
```

request body 格式说明：

| 数据表示内容 | 字段  | 数据类型        | 描述及要求                 |
| :----------- | :---- | :-------------- | :------------------------- |
| 登出时间     | at    | date ISO string | 数据采集的原始时间         |
| 实时车辆数据 | items | [object]        | 车辆实时数据, 具体定义见下 |

##### 1.3.1 整车数据 (type=VEHICLE)

```js
{
  type: "VEHICLE",
  status: "ON",
  chargeStatus: "UNCHARGED",
  mode: "ELECTRIC",
  speed: 0,
  mileage: 178407.5,
  voltage: 570.5,
  current: -31,
  soc: 0.57,
  dcStatus: "OFF",
  shift: "D",
  resistance: 16822,
  aptv: 0,
  brake: 0.01,
}
```

数据说明:

| 字段         | 数据内容       | 类型          | 有效值范围   | 分辨率  | 说明                                                                                               |
| :----------- | :------------- | :------------ | :----------- | :------ | :------------------------------------------------------------------------------------------------- |
| status       | 车辆状态       | string enum   |              |         | "ON": 启动, "OFF": 熄火, "OTHER": 其他                                                             |
| chargeStatus | 充电状态       | string enum   |              |         | "PARK_CHARGING": 停车充电, "MOVE_CHARGING": 行驶充电, "UNCHARGED": 未充电状态, "COMPETE": 充电完成 |
| mode         | 运行模式       | string enum   |              |         | "ELECTRIC": 电动, "MIXED": 混动, "FUEL": 燃油                                                      |
| speed        | 车速           | float         | 0 ~220       | 0.1km/h |
| mileage      | 累计里程       | float         | 0 ～ 9999999 | 0.1km   |
| voltage      | 总电压         | float         | 0 ～ 10000   | 0.1V    |
| current      | 总电流         | float 0~20000 | 0.1A         |
| soc          | SOC            | float         | 0~1          | 0.01    |
| dcStatus     | DC-OC 状态     | string        |              |         | "ON":工作, "OFF": 断开                                                                             |
| shift        | 档位           | enum          |              |         |
| resistance   | 绝缘电阻       | integer       | 0~60000      | 1kΩ     |
| aptv         | 加速踏板行程值 | float         | 0~1          | 0.01    |
| break        | 制动踏板状态   | float         | 0~1          | 0.01    |                                                                                                    |

##### 1.3.2 驱动电机数据 (type=MOTOR)

```js
{
  type: "MOTOR",
  count: 1,
  motors: [
    {
      no: 1,
      status: "READY",
      controlTemp: 0,
      speed: 0,
      torque: 0,
      temp: 0,
      voltage: 0,
      current: 0,
    },
  ],
}
```

数据说明

| 字段   | 数据内容             | 类型     | 有效值范围 |
| :----- | :------------------- | :------- | :--------- |
| count  | 驱动电机个数         | integer  | 1~253      |
| motors | 驱动电机总成信息列表 | [object] |            |

每个驱动电机数据说明:

| 字段        | 数据内容               | 类型        | 有效值范围 | 分辨率 | 说明                                                                      |
| :---------- | :--------------------- | :---------- | :--------- | :----- | :------------------------------------------------------------------------ |
| no          | 驱动电机序号           | integer     | 1~253      |        |                                                                           |
| status      | 驱动电机状态           | enum string |            |        | "CONSUMPTION": 耗电， "GENERATION": 发电， "OFF": 关闭状态, "READY": 准备 |
| controlTemp | 驱动电机控制器温度     | float       | 0~250      | 1℃     |                                                                           |
| speed       | 驱动电机转速           | integer     | 0~65531    | 1r/min |                                                                           |
| torque      | 驱动电机转矩           | integer     | 0~65531    | 0.1N·m |                                                                           |
| temp        | 驱动电机温度           | integer     | 0~250      | 1℃     |                                                                           |
| voltage     | 电机控制器输入电压     | integer     | 0~60000    | 0.1V   |                                                                           |
| current     | 电机控制器直线母线电流 | integer     | 0~20000    | 0.1A   |                                                                           |

##### 1.3.3 车辆位置数据 (type=LOCATION)

原始日志中的位置坐标系使用的是 WGS－84 原始坐标系, 在解析成 record 之后会转变成 GCJ－02 坐标系

```js
{
  type: "LOCATION",
  state: 0,
  lng: 121.4482,
  lat: 31.25105
}
```

数据说明:

| 字段  | 数据内容 | 类型    | 说明                                         |
| :---- | :------- | :------ | :------------------------------------------- |
| state | 定义状态 | integer | 0：有效定位；1：无效定位                     |
| lng   | 经度     | float   | 以度为单位的纬度值乘以 10^6,精度到万分之一度 |
| lat   | 纬度     | float   | 以度为单位的纬度值乘以 10^6,精度到万分之一度 |

##### 1.3.4 极值数据 (type=EXTREME)

```js
{
  type: "EXTREME",
  maxVoltageSubSysNo: 1,
  maxVoltageSingNo: 63,
  maxVoltage: 3.263,
  minVoltageSubSysNo: 1,
  minVoltageSingNo: 91,
  minVoltage: 3.25,
  maxNtcSubSysNo: 1,
  maxNtcNo: 2,
  maxNtc: 30,
  minNtcSubSysNo: 1,
  minNtcNo: 78,
  minNtc: 24,
}
```

数据说明:

| 字段               | 数据内容             | 类型    | 有效值范围 | 分辨率 | 说明 |
| :----------------- | :------------------- | :------ | :--------- | :----- | :--- |
| maxVoltageSubSysNo | 最高电压电池子系统号 | integer | 1~250      |        |      |
| maxVoltageSingNo   | 最高电压电池单体代号 | integer | 1~250      |        |      |
| maxVoltage         | 电池单体电压最高值   | float   | 1~15000    | 0.001V |      |
| minVoltageSubSysNo | 最低电压电池子系代号 | integer | 1~250      |        |      |
| minVoltageSingNo   | 最低电压电池单体代号 | integer | 1~250      |        |      |
| minVoltage         | 电池单体最低电压值   | float   | 1~15000    | 0.001V |      |
| maxNtcSubSysNo     | 最高温度子系统号     | integer | 1~250      |        |      |
| maxNtcNo           | 最高温度探针序号     | integer | 1~250      |        |      |
| maxNtc             | 最高温度值           | integer | 1~250      | 1℃     |      |
| minNtcSubSysNo     | 最低温度子系统号     | integer | 1~250      |        |      |
| minNtcNo           | 最低温度探针序号     | integer | 1~250      |        |      |
| minNtc             | 最低温度值           | integer | 1~250      | 1℃     |      |

##### 1.3.5 报警数据 (type=ALARM)

```js
{
  type: "ALARM",
  maxLevel: 0,
  uas: {
    ressChargeOver: 0,
    motorTemp: 0,
    highVolMuteStatus: 0,
    motorControlTemp: 0,
    dcdcStatus: 0,
    brake: 0,
    dcdcTemp: 0,
    insulation: 0,

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.7.1"></a>
## [0.7.1](https://github.com/36node/shanghaibus-messenger/compare/v0.7.0...v0.7.1) (2019-09-12)



<a name="0.7.0"></a>
# [0.7.0](https://github.com/36node/shanghaibus-messenger/compare/v0.6.0...v0.7.0) (2019-09-12)


### Bug Fixes

* ci use node 10 ([d37ae3f](https://github.com/36node/shanghaibus-messenger/commit/d37ae3f))
* elint warning ([1a85964](https://github.com/36node/shanghaibus-messenger/commit/1a85964))
* github security warning ([c3b499d](https://github.com/36node/shanghaibus-messenger/commit/c3b499d))
* node_kafka version ([6a74e1a](https://github.com/36node/shanghaibus-messenger/commit/6a74e1a))
* remove clg ([982e713](https://github.com/36node/shanghaibus-messenger/commit/982e713))
* try to update rdkafka version ([4984320](https://github.com/36node/shanghaibus-messenger/commit/4984320))


### Features

* add adas2,chargingBow,energy parse ([89e879c](https://github.com/36node/shanghaibus-messenger/commit/89e879c))



<a name="0.6.0"></a>
# [0.6.0](https://github.com/36node/shanghaibus-messenger/compare/v0.5.8...v0.6.0) (2019-03-14)


### Features

* use gcoord transform loaction on parse log ([ed92264](https://github.com/36node/shanghaibus-messenger/commit/ed92264))



<a name="0.5.8"></a>
## [0.5.8](https://github.com/36node/shanghaibus-messenger/compare/v0.5.7...v0.5.8) (2019-03-07)


### Bug Fixes

* handle type ADAS ([#10](https://github.com/36node/shanghaibus-messenger/issues/10)) ([422a728](https://github.com/36node/shanghaibus-messenger/commit/422a728))
* readme and typedef ([#10](https://github.com/36node/shanghaibus-messenger/issues/10)) ([9d7c3b0](https://github.com/36node/shanghaibus-messenger/commit/9d7c3b0))



<a name="0.5.7"></a>
## [0.5.7](https://github.com/36node/shanghaibus-messenger/compare/v0.5.6...v0.5.7) (2019-02-26)


### Bug Fixes

* keyPosition typo ([de6e213](https://github.com/36node/shanghaibus-messenger/commit/de6e213))



<a name="0.5.6"></a>
## [0.5.6](https://github.com/36node/shanghaibus-messenger/compare/v0.5.5...v0.5.6) (2019-01-30)


### Bug Fixes

* index.d.ts ([89020da](https://github.com/36node/shanghaibus-messenger/commit/89020da))



<a name="0.5.5"></a>
## [0.5.5](https://github.com/36node/shanghaibus-messenger/compare/v0.5.4...v0.5.5) (2019-01-30)


### Bug Fixes

* index.d.ts ([960aebb](https://github.com/36node/shanghaibus-messenger/commit/960aebb))



<a name="0.5.4"></a>
## [0.5.4](https://github.com/36node/shanghaibus-messenger/compare/v0.5.3...v0.5.4) (2019-01-30)


### Bug Fixes

* index.d.ts ([b36c648](https://github.com/36node/shanghaibus-messenger/commit/b36c648))



<a name="0.5.3"></a>
## [0.5.3](https://github.com/36node/shanghaibus-messenger/compare/v0.5.2...v0.5.3) (2019-01-28)


### Bug Fixes

* add typings dir in package.json ([ad59dbd](https://github.com/36node/shanghaibus-messenger/commit/ad59dbd))



<a name="0.5.2"></a>
## [0.5.2](https://github.com/36node/shanghaibus-messenger/compare/v0.5.0...v0.5.2) (2019-01-28)


### Bug Fixes

* remove export default ([bf8a049](https://github.com/36node/shanghaibus-messenger/commit/bf8a049))



<a name="0.5.1"></a>
## [0.5.1](https://github.com/36node/shanghaibus-messenger/compare/v0.5.0...v0.5.1) (2019-01-28)


### Bug Fixes

* remove export default ([bf8a049](https://github.com/36node/shanghaibus-messenger/commit/bf8a049))



<a name="0.5.0"></a>
# [0.5.0](https://github.com/36node/shanghaibus-messenger/compare/v0.4.0...v0.5.0) (2019-01-27)


### Features

* add typings ([30b6a42](https://github.com/36node/shanghaibus-messenger/commit/30b6a42))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/36node/shanghaibus-messenger/compare/v0.3.3...v0.4.0) (2019-01-26)


### Features

* add kafka_message to parsed log ([5388f62](https://github.com/36node/shanghaibus-messenger/commit/5388f62))



<a name="0.3.3"></a>
## [0.3.3](https://github.com/36node/shanghaibus-messenger/compare/v0.3.2...v0.3.3) (2019-01-17)


### Bug Fixes

* support large maxLevel ([56dc515](https://github.com/36node/shanghaibus-messenger/commit/56dc515))



<a name="0.3.2"></a>
## [0.3.2](https://github.com/36node/shanghaibus-messenger/compare/v0.3.1...v0.3.2) (2019-01-17)


### Bug Fixes

* alarm parse ([b32a43b](https://github.com/36node/shanghaibus-messenger/commit/b32a43b))



<a name="0.3.1"></a>
## [0.3.1](https://github.com/36node/shanghaibus-messenger/compare/v0.3.0...v0.3.1) (2019-01-16)


### Bug Fixes

* cost and partial maybe false or 0 ([f163b1d](https://github.com/36node/shanghaibus-messenger/commit/f163b1d))



<a name="0.3.0"></a>
# [0.3.0](https://github.com/36node/shanghaibus-messenger/compare/v0.2.3...v0.3.0) (2019-01-16)


### Features

* add data,origin,cost,partial to parsed log ([e33aa45](https://github.com/36node/shanghaibus-messenger/commit/e33aa45))



<a name="0.2.3"></a>
## [0.2.3](https://github.com/36node/shanghaibus-messenger/compare/v0.2.2...v0.2.3) (2019-01-10)



<a name="0.2.2"></a>
## [0.2.2](https://github.com/36node/shanghaibus-messenger/compare/v0.2.1...v0.2.2) (2019-01-10)



<a name="0.2.1"></a>
## [0.2.1](https://github.com/36node/shanghaibus-messenger/compare/v0.2.0...v0.2.1) (2019-01-10)



<a name="0.2.0"></a>
# [0.2.0](https://github.com/36node/shanghaibus-messenger/compare/v0.1.3...v0.2.0) (2019-01-10)


### Features

* add recordId of parsed log ([c5c8111](https://github.com/36node/shanghaibus-messenger/commit/c5c8111))



<a name="0.1.3"></a>
## [0.1.3](https://github.com/36node/shanghaibus-messenger/compare/v0.1.2...v0.1.3) (2019-01-04)


### Bug Fixes

* record vehicle bug ([e94a345](https://github.com/36node/shanghaibus-messenger/commit/e94a345))



<a name="0.1.2"></a>
## [0.1.2](https://github.com/36node/shanghaibus-messenger/compare/v0.1.0...v0.1.2) (2019-01-04)


### Bug Fixes


src/parser.js
/**
 * 日志的解析
 */
import Ajv from "ajv";
import { logSchema, rdbDataSchema } from "./schemas";
import { pick, uniq } from "lodash";
import { ALARM_FLAGS } from "./constants";
import moment from "moment";
import { isNil, padStart } from "lodash";
import gcoord from "gcoord";

const ajv = new Ajv();

/**
 * 验证message格式是否正确，如果不正确，则抛弃
 * @param message 信息
 */
function validateLog(message) {
  const valid = ajv.validate(logSchema, message);
  if (!valid) {
    const error = new Error("LogSchemaError");
    error.name = "LogSchemaError";
    error.ajv = ajv.errors;
    throw error;
  }
}

export function parseRecord(vin, body) {
  const { items = [], at } = body;
  let record = { id: vin, alarmLevel: 0, at };
  for (let item of items) {
    switch (item.type) {
      case "VEHICLE": {
        record.vehicle = {
          ...pick(item, [
            "status",
            "chargeStatus",
            "mode",
            "speed",
            "mileage",
            "voltage",
            "current",
            "soc",
            "dcStatus",
            "shift",
            "resistance",
            "aptv",
            "brake",
          ]),
        };
        break;
      }
      case "MOTOR": {
        const { motors = [] } = item;
        record.motors = motors;
        break;
      }
      case "LOCATION": {
        const [lng, lat] = gcoord.transform(
          [item.lng, item.lat],
          gcoord.WGS84,
          gcoord.GCJ02
        );
        record.location = { lng, lat };
        break;
      }
      case "EXTREME": {
        record.extreme = pick(item, [
          "maxVoltageSubSysNo",
          "maxVoltageSingNo",
          "maxVoltage",
          "minVoltageSubSysNo",
          "minVoltageSingNo",
          "minVoltage",
          "maxNtcSubSysNo",
          "maxNtcNo",
          "maxNtc",
          "minNtcSubSysNo",
          "minNtcNo",
          "minNtc",
        ]);
        break;
      }
      case "ALARM": {
        const { maxLevel = 0, uas = {} } = item;
        const codes = [];
        record.alarmLevel = maxLevel;
        // 处理uas的警报
        // uas 解析出的警报等级不是根据 maxLevel 定义的，而是 找出当前标志位对应的警报码，再由警报表中查询出来等级
        if (maxLevel > 0) {
          // 有效的标志位
          const tags = Object.keys(uas).filter(k => uas[k] && uas[k] > 0);

          const level = Math.min(3, maxLevel);

          // 解析为警报码
          codes.push(
            ...tags.map(t => {
              const defaultCode =
                "000000" + padStart(level.toString(16), 2, "0");
              if (!ALARM_FLAGS[t]) return defaultCode;

              const code = ALARM_FLAGS[t][level - 1];
              if (!code) return defaultCode;

              return code.toString(16);
            })
          );
        }

        // 处理 list 的警报
        const {
          ressList = [],
          mortorList = [],
          engineList = [],
          otherList = [],
        } = item;

        const toCode = (alarm = {}) => {
          const { type = 0, code = 0, level = 1 } = alarm;

          return (
            padStart(type.toString(16), 2, "0") +
            padStart(code.toString(16), 4, "0") +
            padStart(level.toString(16), 2, "0")
          );
        };

        [ressList, mortorList, engineList, otherList].forEach(l => {
          codes.push(...l.map(a => toCode(a)));
        });

        record.alarms = uniq(codes);

        break;
      }
      case "CUSTOM_EXT": {
        record.customExt = pick(item, [
          "pressure1", // 气压1
          "pressure2", // 气压2
          "batteryVoltage", // 蓄电池电压
          "dcov", // DCDC输出电压
          "dcoc", // DCDC输出电流
          "dcTemp", // DCDC散热器温度
          "acTemp", // DCAC散热器温度
          "lftp", // 左前轮胎压力
          "lftt", // 左前轮胎温度
          "rftp", // 右前轮胎压力
          "rftt", // 右前轮胎温度
          "lr1tp", // 左后 1 轮胎压力
          "lr1tt", // 左后 1 轮胎温度
          "lr2tp", // 左后 2 轮胎压力
          "lr2tt", // 左后 2 轮胎温度
          "rr1tp", // 右后 1 轮胎压力
          "rr1tt", // 右后 1 轮胎温度
          "rr2tp", // 右后 2 轮胎压力
          "rr2tt", // 右后 2 轮胎温度
          "cv", // 充电电压
          "rc", // 充电电流
          "cp", // 充电电量
          "totalCharge", // 累积充电电量
          "totalDischarge", // 累积放电电量
          "instantPower", // 瞬时电耗
          "bpiRes", // 电池正绝缘电阻
          "bniRes", // 电池负绝缘电阻
          "apTemp", // 气泵扇热器温度
          "motorContTemp", // 电机控制器温度
          "airMode", // 空调模式", 关闭, 进风, 制热, 制冷
          "airTemp", // 空调设定温度
          "insideTemp", // 车厢内实际温度
          "outsideTemp", // 车外温度
          "middleDoorStatus", // 中门状态", 关闭, 开, 异常
          "frontDoorStatus", // 前门状态", 关闭, 开, 异常
          "handbrakeStatus", // 手刹状态", 松, 刹, 异常
          "keyPosition", // 钥匙位置
        ]);
        break;
      }
      case "TEN_SECONDS": {
        const { datas = [] } = item;
        const { tens = [] } = record;
        record.tens = [...tens, ...datas];

        break;
      }
      case "ADAS": {
        const { datas = [] } = item;
        const { adas = [] } = record;
        record.adas = [...adas, ...datas];
        break;
      }
      case "ADAS2": {
        const { datas = [] } = item;
        const { adas2 = [] } = record;
        record.adas2 = [...adas2, ...datas];
        break;
      }
      case "CHARGING_BOW": {
        const { datas = [] } = item;
        const { chargingBow = [] } = record;
        record.chargingBow = [...chargingBow, ...datas];
        break;
      }
      case "ENERGY": {
        record.energy = pick(item, [
          "length",
          "defrostTotal",
          "defrost",
          "airTotal",
          "air",
          "motorTotal",
          "motor",
          "turnTotal",
          "turn",
          "compressorTotal",
          "compressor",
          "dcdcTotal",
          "dcdc",
          "heatingReverseTotal",
          "heatingTotal",
          "heatingReverse",
          "heating",
          "reverseTotal",
        ]);
        break;
      }
      default:
        break;
    }
  }
  return record;
}

function handleCommand(request) {
  const { command, vin, body } = request;

  switch (command) {
    case "REISSUE_REPORT":
    case "REALTIME_REPORT":
      const record = parseRecord(vin, body);
      return {
        type: "RDB_DATA",
        vin: vin,
        command: command,
        payload: record,
      };
    case "VEHICLE_LOGIN":
    case "VEHICLE_LOGOUT":
    case "HEARTBEAT":
      return {
        type: "RDB_DATA",
        vin: vin,
        command: command,
        payload: body,
      };
    default:
      throw new Error("Not support request command:", command);
  }
}

/**
 * 处理 info 日志
 * @param {*} log
 */
function handleInfo(log) {
  // 如果是 rdb data
  const valid = ajv.validate(rdbDataSchema, log);
  if (!valid) {
    return {
      type: "INVALID_LOG",
      payload: log,
      error: ajv.errors,
    };
  } else {
    return handleCommand(log.request);
  }
}

export function handleKafkaData(kafkaData) {
  const logStr = kafkaData.value.toString() || "";
  const parsedLog = parseLog(logStr);
  // 保存原始日志
  if (parsedLog) parsedLog.kafka_message = logStr;
  return parsedLog;
}

/**
 * 解析log字符串
 * @param {string} logStr
 */
export function parseLog(logStr) {
  try {
    const message = JSON.parse(logStr) || {};
    const { log } = message;
    if (!log && typeof log === "string") {
      throw new Error(
        "Message must contains log properity, which must be a string"
      );
    }
    const logObject = JSON.parse(log);
    validateLog(logObject);

    const {
      level,
      time,
      session,
      seq,
      cost,
      data,
      origin,
      partial,
    } = logObject;

    let finalLog = {};

    switch (level) {
      case 30:
        finalLog = handleInfo(logObject);
        break;
      case 50:
        finalLog = {
          type: "REQUEST_ERROR",
          payload: logStr,
        };
        break;
      default:
        finalLog = {
          type: "INVALID_LOG",
          payload: logStr,
          error: "Non support log type!",
        };
        break;
    }

    finalLog.reportedAt = moment(time).toISOString();

    if (session && seq !== undefined) {
      finalLog.recordId = session + "_" + seq;
    }

    if (!isNil(cost)) {
      finalLog.cost = cost;
    }

    if (!isNil(data)) {
      finalLog.data = data;
    }

    if (!isNil(partial)) {
      finalLog.partial = partial;
    }

    if (!isNil(origin)) {
      finalLog.origin = origin;
    }

    return finalLog;
  } catch (error) {
    // console.log(error.name);

    // json 解析错误，直接抛弃日志
    if (error.name === "SyntaxError") {
      return;
    }

    // 不符合 logSchema的 也直接抛弃日志
    if (error.name === "LogSchemaError") {
      return;
    }

    return {
      type: "INVALID_LOG",
      payload: logStr,
      error: error.toString(),
    };
  }
}
